Wire contexts through a couple of subsystems

cpuguy83 opened the pull request:
This plumbs through a context within `plugin/` and `libcontainerd/`.
In some cases this touched some higher-level things.

Note: I did not change `pkg/plugins` or `pkg/plugingetter` because this is a *much* larger change (deceptively so) which won't even compile until some swarmkit changes are made (fun times here: https://github.com/cpuguy83/docker/commit/9c25233a2f24c6b0b84790af628044fe6edd1aa8).

Basically taking a bottom-up approach here so the underlying subsystems are ready to take a context when ready.

cpuguy83 committed:
plugin: Plumb context through

This required some additional changes in the container event handler,
hence why this is not touching only the plugin system.

This did not change the `PluginGetter` interface which is defined in
pkg/plugins as this is another semi-large change across different
packages.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

thaJeztah reviewed (dismissed):
LGTM

thaJeztah commented:
oh, scratch that; looks like a build is failing;


```
8 65.69 # github.com/docker/docker/plugin
#38 65.69 plugin/manager.go:54:29: too many arguments in call to pm.restore
#38 65.69 	have (context.Context, *"github.com/docker/docker/plugin/v2".Plugin, *controller)
#38 65.69 	want (*"github.com/docker/docker/plugin/v2".Plugin, *controller)
#38 65.69 plugin/manager.go:150:24: too many arguments in call to pm.enable
#38 65.69 	have (context.Context, *"github.com/docker/docker/plugin/v2".Plugin, *controller, bool)
#38 65.69 	want (*"github.com/docker/docker/plugin/v2".Plugin, *controller, bool)
#38 65.69 plugin/manager.go:244:36: too many arguments in call to pm.enable
#38 65.69 	have (context.Context, *"github.com/docker/docker/plugin/v2".Plugin, *controller, bool)
#38 65.69 	want (*"github.com/docker/docker/plugin/v2".Plugin, *controller, bool)
#38 74.47 # github.com/docker/docker/libcontainerd/local
#38 74.47 libcontainerd/local/local_windows.go:170:70: not enough arguments in call to c.backend.ProcessEvent
#38 74.47 	have (string, "github.com/docker/docker/libcontainerd/types".EventType, "github.com/docker/docker/libcontainerd/types".EventInfo)
#38 74.47 	want (context.Context, string, "github.com/docker/docker/libcontainerd/types".EventType, "github.com/docker/docker/libcontainerd/types".EventInfo)
#38 74.47 libcontainerd/local/local_windows.go:506:89: not enough arguments in call to ctr.client.backend.ProcessEvent
#38 74.47 	have (string, "github.com/docker/docker/libcontainerd/types".EventType, "github.com/docker/docker/libcontainerd/types".EventInfo)
#38 74.47 	want (context.Context, string, "github.com/docker/docker/libcontainerd/types".EventType, "github.com/docker/docker/libcontainerd/types".EventInfo)
#38 74.47 libcontainerd/local/local_windows.go:662:89: not enough arguments in call to t.ctr.client.backend.ProcessEvent
#38 74.47 	have (string, "github.com/docker/docker/libcontainerd/types".EventType, "github.com/docker/docker/libcontainerd/types".EventInfo)
#38 74.47 	want (context.Context, string, "github.com/docker/docker/libcontainerd/types".EventType, "github.com/docker/docker/libcontainerd/types".EventInfo)
#38 74.47 libcontainerd/local/local_windows.go:670:90: not enough arguments in call to t.ctr.client.backend.ProcessEvent
#38 74.47 	have (string, "github.com/docker/docker/libcontainerd/types".EventType, "github.com/docker/docker/libcontainerd/types".EventInfo)
#38 74.47 	want (context.Context, string, "github.com/docker/docker/libcontainerd/types".EventType, "github.com/docker/docker/libcontainerd/types".EventInfo)
#38 74.47 libcontainerd/local/local_windows.go:794:86: not enough arguments in call to t.ctr.client.backend.ProcessEvent
#38 74.47 	have (string, "github.com/docker/docker/libcontainerd/types".EventType, "github.com/docker/docker/libcontainerd/types".EventInfo)
#38 74.47 	want (context.Context, string, "github.com/docker/docker/libcontainerd/types".EventType, "github.com/docker/docker/libcontainerd/types".EventInfo)
#38 74.47 libcontainerd/local/local_windows.go:835:87: not enough arguments in call to t.ctr.client.backend.ProcessEvent
#38 74.47 	have (string, "github.com/docker/docker/libcontainerd/types".EventType, "github.com/docker/docker/libcontainerd/types".EventInfo)
#38 74.47 	want (context.Context, string, "github.com/docker/docker/libcontainerd/types".EventType, "github.com/docker/docker/libcontainerd/types".EventInfo)
#38 74.47 libcontainerd/local/local_windows.go:1174:84: not enough arguments in call to p.ctr.client.backend.ProcessEvent
#38 74.47 	have (string, "github.com/docker/docker/libcontainerd/types".EventType, "github.com/docker/docker/libcontainerd/types".EventInfo)
#38 74.47 	want (context.Context, string, "github.com/docker/docker/libcontainerd/types".EventType, "github.com/docker/docker/libcontainerd/types".EventInfo)
#38 75.05 # github.com/docker/docker/libcontainerd/remote
#38 75.05 libcontainerd/remote/client.go:223:52: too many arguments in call to newFIFOSet
#38 75.05 	have (context.Context, string, string, bool, bool)
#38 75.05 	want (string, string, bool, bool)
#38 75.05 libcontainerd/remote/client.go:276:87: too many arguments in call to newFIFOSet
#38 75.05 	have (context.Context, string, string, bool, bool)
#38 75.05 	want (string, string, bool, bool)
```

cpuguy83 committed:
libcontainerd: wire through contexts internally

A couple of places we were creating a context when we should be using
the inheritted context.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
Ah the "other" platform 😄 . Should be good now.

thaJeztah commented:
Missed 2 😅

```
libcontainerd/remote/client.go:223:47: not enough arguments in call to newFIFOSet
136.3 	have (string, string, bool, bool)
136.3 	want (context.Context, string, string, bool, bool)
136.3 libcontainerd/remote/client.go:276:82: not enough arguments in call to newFIFOSet
136.3 	have (string, string, bool, bool)
136.3 	want (context.Context, string, string, bool, bool)

cpuguy83 commented:
WAT? Compiles here... 🤔

cpuguy83 commented:
Ok, now it's good realzies this time.

cpuguy83 commented:
Looks like I missed a timeout context masking a parent context, so the `process.Delete` was failing due the context being timed out.
Fixed that... will see what the tests say.

thaJeztah commented:
Saw the diff (and the typo fix 😂), thanks! 🤞

thaJeztah reviewed (approved):
LGTM

vvoland commented on libcontainerd/local/local_windows.go:
```diff
@@ -791,7 +791,7 @@ func (t *task) Pause(_ context.Context) error {
 
 	t.ctr.client.eventQ.Append(t.ctr.id, func() {
-		err := t.ctr.client.backend.ProcessEvent(t.ctr.id, libcontainerdtypes.EventPaused, libcontainerdtypes.EventInfo{
+		err := t.ctr.client.backend.ProcessEvent(ctx, t.ctr.id, libcontainerdtypes.EventPaused, libcontainerdtypes.EventInfo{
```
Or we can just leave leave it as `context.TODO()` and replace with `context.WithoutCancel` when we move to `go1.21`.

cpuguy83 replied:
Cool updated this to use `context.TODO` with a comment explaining why.

vvoland commented on libcontainerd/local/local_windows.go:
```diff
@@ -832,7 +838,7 @@ func (t *task) Resume(ctx context.Context) error {
 
 	t.ctr.client.eventQ.Append(t.ctr.id, func() {
-		err := t.ctr.client.backend.ProcessEvent(t.ctr.id, libcontainerdtypes.EventResumed, libcontainerdtypes.EventInfo{
+		err := t.ctr.client.backend.ProcessEvent(ctx, t.ctr.id, libcontainerdtypes.EventResumed, libcontainerdtypes.EventInfo{
```
This one also needs a `context.TODO()`.

vvoland commented on libcontainerd/remote/client.go:
```diff
@@ -220,9 +220,9 @@ func (c *container) Start(ctx context.Context, checkpointDir string, withStdin b
 		func(id string) (cio.IO, error) {
-			fifos := newFIFOSet(bundle, id, withStdin, spec.Process.Terminal)
+			fifos := newFIFOSet(ctx, bundle, id, withStdin, spec.Process.Terminal)
 
-			rio, err = c.createIO(fifos, stdinCloseSync, attachStdio)
+			rio, err = c.createIO(ctx, fifos, stdinCloseSync, attachStdio)
```
Was wondering how about these two.

Looked at the current implementation and this is called right in `NewTask`: https://github.com/containerd/containerd/blob/40f26543bdc27cbe8b058ac082e91c5832bb1c41/container.go#L216

Looking at the usage, it doesn't seem like this would change, so I think it's fine.

vvoland commented on daemon/monitor.go:
```diff
@@ -206,7 +206,7 @@ func (daemon *Daemon) ProcessEvent(id string, e libcontainerdtypes.EventType, ei
 			if execConfig.Process != nil {
 				go func() {
-					if _, err := execConfig.Process.Delete(context.Background()); err != nil {
+					if _, err := execConfig.Process.Delete(ctx); err != nil {
```
I think `context.Background` was an explicit choice there, so looks like another potential `context.WithoutCancel`.

vvoland commented on daemon/monitor.go:
```diff
@@ -39,7 +39,7 @@ func (daemon *Daemon) handleContainerExit(c *container.Container, e *libcontaine
 	tsk, ok := c.Task()
 	if ok {
-		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+		ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
```
Shouldn't this context cancellation be separate from the `ctx` passed to this function?
Otherwise, if `handleContainerExit` gets cancelled, the context passed to `tsk.Delete` will already be "done" and the task won't be deleted.

vvoland commented on libcontainerd/local/local_windows.go:
```diff
@@ -167,7 +167,7 @@ func (c *client) NewContainer(_ context.Context, id string, spec *specs.Spec, sh
 				"event":     libcontainerdtypes.EventCreate,
 			}).Info("sending event")
-			err := c.backend.ProcessEvent(id, libcontainerdtypes.EventCreate, ei)
+			err := c.backend.ProcessEvent(ctx, id, libcontainerdtypes.EventCreate, ei)
```
This `ctx` is also captured in a closure that will be executed asynchronously.

vvoland commented on libcontainerd/local/local_windows.go:
```diff
@@ -503,7 +503,7 @@ func (ctr *container) Start(_ context.Context, _ string, withStdin bool, attachS
 			"event-info": ei,
 		}).Info("sending event")
-		err := ctr.client.backend.ProcessEvent(ei.ContainerID, libcontainerdtypes.EventStart, ei)
+		err := ctr.client.backend.ProcessEvent(ctx, ei.ContainerID, libcontainerdtypes.EventStart, ei)
```
And this one too.

vvoland commented on libcontainerd/local/local_windows.go:
```diff
@@ -659,15 +659,15 @@ func (t *task) Exec(ctx context.Context, processID string, spec *specs.Process,
 			"event-info": ei,
 		}).Info("sending event")
-		err := t.ctr.client.backend.ProcessEvent(t.ctr.id, libcontainerdtypes.EventExecAdded, ei)
+		err := t.ctr.client.backend.ProcessEvent(ctx, t.ctr.id, libcontainerdtypes.EventExecAdded, ei)
```
This also needs to be non-cancellable.

vvoland commented on libcontainerd/local/local_windows.go:
```diff
@@ -659,15 +659,15 @@ func (t *task) Exec(ctx context.Context, processID string, spec *specs.Process,
 			}).Error("failed to process event")
 		}
-		err = t.ctr.client.backend.ProcessEvent(t.ctr.id, libcontainerdtypes.EventExecStarted, ei)
+		err = t.ctr.client.backend.ProcessEvent(ctx, t.ctr.id, libcontainerdtypes.EventExecStarted, ei)
```
And this one.

vvoland commented on libcontainerd/remote/client.go:
```diff
@@ -571,7 +571,7 @@ func closeStdin(ctx context.Context, p containerd.Process) error {
 func (c *client) processEvent(ctx context.Context, et libcontainerdtypes.EventType, ei libcontainerdtypes.EventInfo) {
 	c.eventQ.Append(ei.ContainerID, func() {
-		err := c.backend.ProcessEvent(ei.ContainerID, et, ei)
+		err := c.backend.ProcessEvent(ctx, ei.ContainerID, et, ei)
```
Wondering about this one too. In principle I think it should also be not-cancellable, but looking at the code, the context here is "the main context" that's tied to the lifetime of the daemon and it will be cancelled when `DaemonCli.start` finishes. But at this point, the whole program has little time left before the main goroutine exits, so the goroutine calling these functions may not have a chance to run this anyway.

cpuguy83 replied:
Would love to hear others thoughts on this.
All these queue processing contexts would be inheriting from the daemon start context.
Does it make sense to try and process events we've already received before letting the daemon shut down?

I think I assumed for all of these the inherited context would only ever be cancelled by daemon shutdown.
